node/src/container_chain_spawner.rs

```diff
@@ -483,32 +510,60 @@ fn handle_update_assignment_state_change(
             if para_id == orchestrator_para_id {
                 call_collate_on = Some(collate_on_tanssi);
             } else {
-                // When we get assigned to a different container chain, only need to call collate_on if it was already
-                // running before
-                if running_chains_before.contains(&para_id) {
-                    let c = state.spawned_container_chains.get(&para_id).expect("container chain was running before so it should exist in spawned_container_chains");
-                    call_collate_on = Some(c.collate_on.clone());
-                }
+                // When we get assigned to a different container chain, we don't need to call collate_on because
+                // we will restart that container chain in collation mode.
+                need_to_restart_current = true;
+            }
+        }
+
+        if let Some(para_id) = state.assigned_para_id {
+            if para_id != orchestrator_para_id && Some(para_id) == next {
+                need_to_restart_next = true;
             }
         }
     }
 
     state.assigned_para_id = current;
     state.next_assigned_para_id = next;
 
-    let chains_to_stop = running_chains_before
+    let mut chains_to_stop: Vec<_> = running_chains_before
         .difference(&running_chains_after)
         .copied()
         .collect();
-    let chains_to_start = running_chains_after
+    let mut chains_to_start: Vec<_> = running_chains_after
         .difference(&running_chains_before)
         .copied()
         .collect();
 
+    if need_to_restart_current {
+        // Force restart of new assigned container chain: if it was running before it was in "syncing mode",
+        // which doesn't use the correct ports, so start it in "collation mode".
+        let id = current.unwrap();
+        if running_chains_before.contains(&id) && !chains_to_stop.contains(&id) {
+            chains_to_stop.push(id);
+        }
+        if !chains_to_start.contains(&id) {
+            chains_to_start.push(id);
+        }
+    }
+
+    if need_to_restart_next {
+        // Handle edge case of going from (2000, 2001) to (2001, 2000). In that case we must restart both chains,
```

Comment:
in the second case, it restarts with random ports?

Reply:
Yes, we stop both of them, wait 10 seconds, and then start the new "syncing" with random ports, and the new "collating" with the good ports (in either order).